### PR TITLE
Improved Angular compatibility

### PR DESCRIPTION
**Pull request by JBBianchi**
In the current stage, `@jsonforms/angular-material` still relies on `@angular/flex-layout`.

This package has been deprecated and isn't compatible with Angular v16+, therefore `@jsonforms/angular-material` would be stuck to Angular v15.

This PR removes the dependency to `@angular/flex-layout` by replacing the attributes with plain CSS instead.

Doing so, `@jsonforms/angular-material` becomes compatible with Angular v16 and v17 (at first sight).

For testing purposes, I forced upgrade the `jsonforms-angular-seed` projects to both v16 and v17:
- https://github.com/JBBianchi/jsonforms-angular-seed/tree/v15-to-v16
- https://github.com/JBBianchi/jsonforms-angular-seed/tree/v16-to-v17

I tried building the packages as submitted in this PR and running them in the above projects and it **seems** to work fine (it displays, I can remove one of the two items and I can update a text field. I didn't test more than that).

Related to
- https://github.com/eclipsesource/jsonforms/issues/2138

### CODE REVIEW

**Comment by JBBianchi** [comment]
(Follow up of #2212)

**Comment by JBBianchi** [comment]
I might have messed up the [pnpm-lock.yaml](https://github.com/eclipsesource/jsonforms/pull/2229/files#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bb) ...

Is there a way for pnpm to use Angular v15 even if the package is compatible with higher versions ? (compile vs usage)

**Comment by gexclaude** [comment]
@JBBianchi really appreciate your efforts!

**Comment by JBBianchi** [comment]
We could actually use Angular v16 for the build process, which will allow to be compliant with #2224 .

Should I add this to this PR as well ? It's basically just changing the pnpm.overrides and removing the `entryComponents` config from the tests so the impact is pretty low.

@sdirix @lucas-koehler WDYT ?

**Comment by lucas-koehler** [comment]
Hi @JBBianchi,
first of all thanks for the PR and your continued effort on improving the angular renderer set :heart: 
I see no harm in using Angular 16 for the build process already. So gladly go ahead with that :)

**Comment by JBBianchi** [comment]
> Hi @JBBianchi, first of all thanks for the PR and your continued effort on improving the angular renderer set ❤️ I see no harm in using Angular 16 for the build process already. So gladly go ahead with that :)

For ng16, I'll hold my horses because the example doesn't seem to run. I'll need to investigate a bit more even though all lights seemed green at first.

**Comment by JBBianchi** [comment]
@lucas-koehler @sdirix the PR should be ready for review.

File: packages/angular-material/src/controls/range.renderer.ts [hunk 1]
```diff
@@ -33,10 +33,11 @@ import { isRangeControl, RankedTester, rankWith } from '@jsonforms/core';
 @Component({
   selector: 'RangeControlRenderer',
   template: `
-    <div fxFlex fxLayout="column" [fxHide]="hidden">
+    <div [ngStyle]="{ display: hidden ? 'none' : '' }" class="range-control">
       <label class="mat-caption" style="color:rgba(0,0,0,.54)">{{
         label
       }}</label>
+      <!-- TODO: The 'tickInterval' property no longer exists -->
```
Comment (lucas-koehler) [inline packages/angular-material/src/controls/range.renderer.ts:40]: Is there a problem with the current solution (using `showTickMarks`). If it currently works fine, I don't think we need this TODO

File: packages/angular-material/src/layouts/group-layout.renderer.ts [hunk 2]
```diff
@@ -34,19 +34,34 @@ import { JsonFormsAngularService } from '@jsonforms/angular';
...
     </mat-card>
   `,
+  styles: [
+    `
+      .group-layout {
+        display: flex;
+        flex-direction: column;
+        gap: 16px;
+      }
+      .group-layout > * {
```
Comment (lucas-koehler) [inline packages/angular-material/src/layouts/group-layout.renderer.ts:60]: Should this really apply to all elements  or only `div`s? I ask because `mat-card-title` did not have `fxFlex` before

File: packages/angular-material/test/group-layout.spec.ts [hunk 3]
```diff
@@ -46,7 +45,7 @@ describe('Group layout', () => {
   beforeEach(() => {
     fixture = beforeEachLayoutTest(GroupLayoutRenderer, {
       declarations: [LayoutChildrenRenderPropsPipe, MatCard, MatCardTitle],
-      imports: [FlexLayoutModule],
+      imports: [],
```
Comment (lucas-koehler) [inline packages/angular-material/test/group-layout.spec.ts:48]: can this be removed now that it's empty?

File: packages/angular-material/test/horizontal-layout.spec.ts [hunk 4]
```diff
@@ -45,7 +44,7 @@ describe('Horizontal layout', () => {
   beforeEach(() => {
     fixture = beforeEachLayoutTest(HorizontalLayoutRenderer, {
       declarations: [LayoutChildrenRenderPropsPipe],
-      imports: [FlexLayoutModule],
+      imports: [],
```
Comment (lucas-koehler) [inline packages/angular-material/test/horizontal-layout.spec.ts:47]: can this be removed now that it's empty?

File: packages/angular-material/tsconfig.spec.json [hunk 5]
```diff
@@ -11,11 +11,5 @@
   "include": [
     "**/*.spec.ts",
     "**/*.d.ts"
-  ],
-  "angularCompilerOptions": {
-    "compilationMode": "partial",
-    "strictMetadataEmit": true,
-    "generateCodeForLibraries": false,
-    "skipTemplateCodegen": true
```
Comment (lucas-koehler) [inline packages/angular-material/tsconfig.spec.json:19]: Why did you remove this? Are these all defaults?

**Review by lucas-koehler — COMMENTED** [review]
Hi thanks again for the PR :heart: 
The changes already look pretty good to me :)
I have some minor comments/questions inline.

File: packages/angular-material/tsconfig.spec.json (hunk 5 above)
Comment (JBBianchi) [reply packages/angular-material/tsconfig.spec.json:19]: I must admit I don't think they are necessary anymore. If I'm not mistaken, the only one still present when generating a lib with the angular cli now is `generateCodeForLibraries` for the production build.

In doubt, I'll add them again.

File: packages/angular-material/test/horizontal-layout.spec.ts (hunk 4 above)
Comment (JBBianchi) [reply packages/angular-material/test/horizontal-layout.spec.ts:47]: sure

File: packages/angular-material/test/group-layout.spec.ts (hunk 3 above)
Comment (JBBianchi) [reply packages/angular-material/test/group-layout.spec.ts:48]: sure

File: packages/angular-material/src/layouts/group-layout.renderer.ts (hunk 2 above)
Comment (JBBianchi) [reply packages/angular-material/src/layouts/group-layout.renderer.ts:60]: I don't think it have an impact but to be on the safe side I'll specify div indeed.

File: packages/angular-material/src/controls/range.renderer.ts (hunk 1 above)
Comment (JBBianchi) [reply packages/angular-material/src/controls/range.renderer.ts:40]: Indeed, makes sense.

File: packages/angular-material/src/layouts/group-layout.renderer.ts [hunk 6]
```diff
@@ -34,19 +34,34 @@ import { JsonFormsAngularService } from '@jsonforms/angular';
...
         "
-        fxFlex
       >
         <jsonforms-outlet [renderProps]="props"></jsonforms-outlet>
       </div>
     </mat-card>
   `,
+  styles: [
+    `
+      .group-layout {
```
Comment (lucas-koehler) [inline packages/angular-material/src/layouts/group-layout.renderer.ts:55]: I think the .group-layout class misses a `padding: 16px;`. Before there was some distance to the borders of the group. This is gone in this PR. The padding in before this was 16px.

**Review by lucas-koehler — CHANGES_REQUESTED** [review]
Hi the code changes LGTM. I had a look at the example. It looked pretty good to me. I only noticed that the group layout is now a bit cramped due to missing padding.

File: packages/angular-material/src/layouts/group-layout.renderer.ts (hunk 6 above)
Comment (JBBianchi) [reply packages/angular-material/src/layouts/group-layout.renderer.ts:55]: It's an artifact from the previous PR but indeed, `.mat-card` had a padding of 16px before. I'll fix it.

**Comment by lucas-koehler** [comment]
@JBBianchi Thanks for the update. Do you know whether the code will still work for Angular 15 when it's built with Angular 16 ?
If yes that's fine for me :) Otherwise, we can also remove compatibility to Angular 15.

**Review by lucas-koehler — CHANGES_REQUESTED** [review]
In the seed, I tested the changes with 15, 16 and 17. The seed did not compile with 15 but worked with 16 and 17. Thus, I suggest removing compatibility to 15. It is end of life in less than half a year anyway.

Also I noticed another bug, the `packages/angular-material/src/controls/range.renderer.ts` does no longer have functioning data binding. I think this might already have been introduced with the Angular 15 support PR when the value and onChange properties were removed in `packages/angular-material/src/controls/range.renderer.ts`

**Comment by JBBianchi** [comment]
> In the seed, I tested the changes with 15, 16 and 17. The seed did not compile with 15 but worked with 16 and 17. Thus, I suggest removing compatibility to 15. It is end of life in less than half a year anyway.
> 
> Also I noticed another bug, the `packages/angular-material/src/controls/range.renderer.ts` does no longer have functioning data binding. I think this might already have been introduced with the Angular 15 support PR when the value and onChange properties were removed in `packages/angular-material/src/controls/range.renderer.ts`

Thanks for testing in the seed v15, that's unfortunate but indeed, support could be dropped.

I'll try to have look during next week if I have some time.

**Comment by JBBianchi** [comment]
It seems indeed that a library built with ng 16 is not compatible with ng 15 because the signature of the `@Input()` directive changed.

For the range control, the binding of the change event was "cleaned out" by mistake in this [commit](https://github.com/JBBianchi/jsonforms/commit/e393bfb46ada53685a46a2a2d01cd9408ad77167).

Next commit will remove compatibility with ng 15 and rebind a the event to the range control.

File: package.json [hunk 7]
```diff
@@ -48,5 +48,28 @@
...
+      "@angular-eslint/eslint-plugin-template": "^16.0.0",
+      "@angular-eslint/schematics": "^16.0.0",
+      "@angular-eslint/template-parser": "^16.0.0",
+      "@angular/compiler": "^16.0.0",
+      "@angular/compiler-cli": "^16.0.0",
+      "@angular/platform-browser-dynamic": "^16.0.0",
+      "@ngtools/webpack": "^16.0.0",
+      "ng-packagr": "^16.0.0",
+      "zone.js": "^0.13.0"
+    }
```
Comment (lucas-koehler) [inline package.json:73]: Do we still need these overrides now that we removed Angular 15 compatibility?

**Review by lucas-koehler — COMMENTED** [review]
Hi @JBBianchi thanks for the updates :) Changes look pretty good to me now. I have just one question regarding the pnpm overrides inline.
I also updated the lock file.

File: package.json (hunk 7 above)
Comment (JBBianchi) [reply package.json:73]: I'm not 100% sure. If I'm not mistaken, as the packages declare a compatibility with Angular 16 and 17 (and maybe 18 in the future), pnpm will install the highest version compatible by default, so v17. So we need the overrides to say "build with v16 even if the lib is compatible with v17".

Comment (sdirix) [reply package.json:73]: Can we use Angular `16` in the `devDependencies` and not `16 || 17`? As we don't bundle Angular the `peerDependencies` version is the important one for the consumer and could still be `16 || 17` even of the `devDependencies` only declare `16`.

Comment (JBBianchi) [reply package.json:73]: @sdirix thanks for your input, that's better indeed.

**Review by lucas-koehler — APPROVED** [review]
Hi @JBBianchi , the PR looks good to me now. I updated all dependencies and added overrides for vue because its latest version doesn't work. I added issue https://github.com/eclipsesource/jsonforms/issues/2234 to fix this later.

Thanks again for your effort on the Angular packages :heart:

File: packages/angular-material/src/other/master-detail/master.ts [hunk 8]
```diff
@@ -140,9 +143,6 @@ export const removeSchemaKeywords = (path: string) => {
         top: 0;
         right: 0;
       }
-      .hide {
-        display: none;
```
Comment (stanisev) [inline packages/angular-material/src/other/master-detail/master.ts:144]: Why is this deleted?

Comment (JBBianchi) [reply packages/angular-material/src/other/master-detail/master.ts:144]: It might be an error when factoring out `@angular/flex-layout` with `[ngStyle]="{ display: hidden ? 'none' : '' }"` instead of `[fxHide]="hidden"`. Do you encounter any problem ?

Comment (stanisev) [reply packages/angular-material/src/other/master-detail/master.ts:144]: Yeah right, that's why I am writing. The delete button is always shown. We created a custom listWithDetail renderer to counter this problem. Is there a better solution? Or can you restore the deleted CSS class?

Comment (JBBianchi) [reply packages/angular-material/src/other/master-detail/master.ts:144]: I'm not a maintainer of this repo but I'd assume you can open an issue explaining the problem and them submit a PR to solve it. Then maintainers can maybe accept the PR and push it in the next release.

Comment (sdirix) [reply packages/angular-material/src/other/master-detail/master.ts:144]: Happy to review and merge a PR

Comment (stanisev) [reply packages/angular-material/src/other/master-detail/master.ts:144]: I've created the issue: https://github.com/eclipsesource/jsonforms/issues/2327
Should you give me permission in order to open a MR?

Comment (lucas-koehler) [reply packages/angular-material/src/other/master-detail/master.ts:144]: @stanisev Thanks for creating the issue :) You can gladly open a MR anytime you want.

Comment (lucas-koehler) [reply packages/angular-material/src/other/master-detail/master.ts:144]: Hi @stanisev , did you try to push the branch directly to this repository? This is not possible because you do not have write permissions. You need to fork the repository and push the branch to your fork. Then, you should be able to open a MR against this repository.
